compiler/rustc_codegen_llvm/src/llvm_util.rs

```diff
@@ -389,36 +383,90 @@ pub fn llvm_global_features(sess: &Session) -> Vec<String> {
-        // features also work on the command line instead of having two
-        // different names when the LLVM name and the Rust name differ.
-        to_llvm_feature(sess, feature).iter().map(|f| format!("{}{}", &s[..1], f)).collect()
-    };
-
     // Features implied by an implicit or explicit `--target`.
-    features.extend(sess.target.features.split(',').flat_map(&filter));
+    features.extend(
+        sess.target
+            .features
+            .split(',')
+            .filter(|v| !v.is_empty() && backend_feature_name(v).is_some())
+            .map(String::from),
+    );
 
     // -Ctarget-features
-    features.extend(sess.opts.cg.target_feature.split(',').flat_map(&filter));
+    let supported_features = supported_target_features(sess);
+    features.extend(
+        sess.opts
+            .cg
+            .target_feature
+            .split(',')
+            .filter_map(|s| {
+                let enable_disable = match s.chars().next() {
+                    None => return None,
+                    Some(c @ '+' | c @ '-') => c,
+                    Some(_) => {
+                        let mut diag = sess.struct_warn(&format!(
+                            "unknown feature specified for `-Ctarget-feature`: `{}`",
+                            s
+                        ));
+                        diag.note("features must begin with a `+` to enable or `-` to disable it");
+                        diag.emit();
+                        return None;
+                    }
+                };
+
+                let feature = backend_feature_name(s)?;
+                // Warn against use of LLVM specific feature names on the CLI.
+                if !supported_features.iter().any(|&(v, _)| v == feature) {
+                    let rust_feature = supported_features.iter().find_map(|&(rust_feature, _)| {
+                        let llvm_features = to_llvm_features(sess, rust_feature);
+                        if llvm_features.contains(&feature)
+                            && !llvm_features.contains(&rust_feature)
+                        {
+                            Some(rust_feature)
+                        } else {
+                            None
+                        }
+                    });
+                    let mut diag = sess.struct_warn(&format!(
+                        "unknown feature specified for `-Ctarget-feature`: `{}`",
+                        feature
+                    ));
+                    if let Some(rust_feature) = rust_feature {
+                        diag.note("it is still passed through to the codegen backend");
+                        diag.help(&format!("did you mean: `{}`", rust_feature));
+                    } else {
+                        diag.note("it is still passed through to the codegen backend");
+                        diag.note("consider filing a feature request");
+                    }
+                    diag.emit();
```

Comment:
```suggestion
                    diag.note("it is still passed through to the codegen backend");
                    if let Some(rust_feature) = rust_feature {
                        diag.help(&format!("you might have meant: `{}`", rust_feature));
                    } else {
                        diag.note("consider filing a feature request");
                    }
                    diag.emit();
```

Not too happy with the "consider filing a feature request", but I have no better idea :-/

Reply:
Hm, is there a reason why this wording makes you concerned? My goal is basically to drive people to describe their use-cases for specific features that they are using so that we could add them to `rustc`.

Reply:
It gives the user something to do, but it is kind of... vague. I'm ok with landing it as is though.